source/views/transportation/bus/bus-view.js

```diff
@@ -1,66 +1,117 @@
 // @flow
 
 import React from 'react'
-import {ScrollView} from 'react-native'
+import {ScrollView, RefreshControl} from 'react-native'
 import type {BusLineType} from './types'
 import {BusLine} from './bus-line'
 import moment from 'moment-timezone'
 import {NoticeView} from '../../components/notice'
 import type {TopLevelViewPropsType} from '../../types'
+import delay from 'delay'
+import {reportNetworkProblem} from '../../../lib/report-network-problem'
 
-import {data as defaultBusLines} from '../../../../docs/bus-times.json'
+import * as defaultData from '../../../../docs/bus-times.json'
 
 const TIMEZONE = 'America/Winnipeg'
 
-export class BusView extends React.PureComponent {
-  static defaultProps = {
-    busLines: defaultBusLines,
-  }
+const GITHUB_URL = 'https://stodevx.github.io/AAO-React-Native/bus-times.json'
 
+export class BusView extends React.PureComponent {
   props: TopLevelViewPropsType & {
-    busLines: BusLineType[],
     line: string,
   }
 
   state = {
+    busLines: defaultData.data,
     intervalId: 0,
+    loading: false,
     now: moment.tz(TIMEZONE),
   }
 
   componentWillMount() {
-    // This updates the screen every second, so that the "next bus" times are
-    // updated without needing to leave and come back.
+    this.fetchData()
+    // This updates the screen every five seconds, so that the "next bus"
+    // times are updated without needing to leave and come back.
     this.setState(() => ({intervalId: setInterval(this.updateTime, 5000)}))
   }
 
   componentWillUnmount() {
     clearTimeout(this.state.intervalId)
   }
 
+  fetchData = async () => {
+    const start = Date.now()
+    this.setState(() => ({loading: true}))
+
+    let {data: busLines} = await fetchJson(GITHUB_URL).catch(err => {
+      reportNetworkProblem(err)
+      return defaultData
+    })
+
+    if (process.env.NODE_ENV === 'development') {
+      busLines = defaultData.data
+    }
+
+    this.updateTime()
+
+    // wait 0.5 seconds – if we let it go at normal speed, it feels broken.
+    const elapsed = start - Date.now()
+    if (elapsed < 500) {
+      await delay(500 - elapsed)
+    }
+
+    this.setState(() => ({busLines, loading: false}))
+  }
+
   updateTime = () => {
     this.setState(() => ({now: moment.tz(TIMEZONE)}))
   }
 
+  refreshTime = async () => {
```

Comment:
This function smells a lot like `fetchData`, can any common behavior be extracted out in a convenient way?

Reply:
I haven't come up with anything.